cli/npm/resolution/graph.rs

```diff
@@ -1,198 +1,348 @@
+    child_pkg_nv: NpmPackageNv,
+  },
+  /// A node that was created during snapshotting and is not being used in any path.
+  SnapshotNodeId(NodeId),
+}
+
+/// A pending resolved identifier used in the graph. At the end of resolution, these
+/// will become fully resolved to an `NpmPackageId`.
+#[derive(Clone)]
+struct ResolvedId {
+  nv: NpmPackageNv,
+  peer_dependencies: Vec<ResolvedIdPeerDep>,
+}
+
+impl ResolvedId {
+  /// Gets a hash of the resolved identifier at this current moment in time.
+  ///
+  /// WARNING: A resolved identifier references a value that could change in
+  /// the future, so this should be used with that in mind.
+  pub fn current_state_hash(&self) -> u64 {
+    let mut hasher = DefaultHasher::new();
+    self.nv.hash(&mut hasher);
+    for dep in &self.peer_dependencies {
+      match dep {
+        ResolvedIdPeerDep::ParentReference {
+          parent,
+          child_pkg_nv,
+        } => {
+          match parent {
+            GraphPathNodeOrRoot::Root(root) => root.hash(&mut hasher),
+            GraphPathNodeOrRoot::Node(node) => node.node_id().hash(&mut hasher),
+          }
+          child_pkg_nv.hash(&mut hasher);
+        }
+        ResolvedIdPeerDep::SnapshotNodeId(node_id) => {
+          node_id.hash(&mut hasher);
+        }
       }
-      maybe_next_node = next_node.previous_node.as_ref();
     }
-    false
+    hasher.finish()
+  }
+}
+
+/// Mappings of node identifiers to resolved identifiers. Each node has exactly
+/// one resolved identifier.
+#[derive(Default)]
+struct ResolvedNodeIds {
+  node_to_resolved_id: HashMap<NodeId, (ResolvedId, u64)>,
+  resolved_to_node_id: HashMap<u64, NodeId>,
+}
+
+impl ResolvedNodeIds {
+  pub fn set(&mut self, node_id: NodeId, resolved_id: ResolvedId) {
+    let resolved_id_hash = resolved_id.current_state_hash();
+    if let Some((_, old_resolved_id_key)) = self
+      .node_to_resolved_id
+      .insert(node_id, (resolved_id, resolved_id_hash))
+    {
+      // ensure the old resolved id key is removed as it might be stale
+      self.resolved_to_node_id.remove(&old_resolved_id_key);
+    }
+    self.resolved_to_node_id.insert(resolved_id_hash, node_id);
   }
 
-  fn id_to_key(id: &NpmPackageNodeId) -> String {
-    format!("{}@{}", id.name, id.version)
+  pub fn get(&self, node_id: NodeId) -> Option<&ResolvedId> {
+    self.node_to_resolved_id.get(&node_id).map(|(id, _)| id)
+  }
+
+  pub fn get_node_id(&self, resolved_id: &ResolvedId) -> Option<NodeId> {
+    self
+      .resolved_to_node_id
+      .get(&resolved_id.current_state_hash())
+      .copied()
   }
 }
 
-/// A memory efficient path of the visited specifiers in the tree.
-#[derive(Default, Clone)]
-struct GraphSpecifierPath {
-  previous_node: Option<Arc<GraphSpecifierPath>>,
-  specifier: String,
+/// A pointer to a specific node in a graph path. The underlying node id
+/// may change as peer dependencies are created.
+#[derive(Clone, Debug)]
+struct NodeIdRef(Arc<Mutex<NodeId>>);
```

Comment:
Btw, I tried to use an `Rc<RefCell<NodeId>>` here, but the lsp code keeps complaining about the future not being send even when I wrap the code that calls this in a `tokio::task::spawn_local(async move { ... }).await.unwrap()`. I don't understand why it doesn't work.